sirepo/job_driver/__init__.py

```diff
@@ -352,6 +352,8 @@ async def _slots_ready(self, op):
             return
         await op.op_slot.alloc('Waiting for another simulation to complete')
         await op.run_dir_slot.alloc('Waiting for access to simulation state')
+        if n in (job.OP_IO,):
```

Comment:
This is right. But I think we now have to change the code for [`destroy_op`](https://github.com/radiasoft/sirepo/blob/4159-download-data-file/sirepo/job_driver/__init__.py#L98). We now want to only free a cpu_slot for certain types of OPS_THAT_NEED_SLOTS (probably need a better name for that too since there are now ops that need an op slot but not a cpu slot).

Hypothetical scenario with a configuration of only one cpu_slot that shows the problem:
1. run starts (job.OP_RUN so it acquires a cpu_slot)
2. io starts (job.OP_IO so it doesn't acquire a cpu_slot)
3.  io finishes, destroy_op is called, it frees the cpu_slot (the one job.OP_RUN acquired), but the run is still running
4. Another run comes in, it is able to acquire a cpu_slot, the machine is now oversubscribed (2 runs happening at the same time when there should only be one). 

I think encapsulating behaviors like this in a test would be nice. Configure the system to only have one cpu_slot. In a thread start an op_run that runs forever. Wait for it to move to running. In another thread start an op_run. Wait for it to become pending. In a third thread start an op_io. Wait for the op_io to finish. The second op_run should still be pending. With the code as it currently I think it will transition to running.

While working on this code I think refactoring 'def _slots_ready()` and some of the constants could be helpful. I think using more named constants and inner functions to separate out the different cases (ops that don't need any slots, sbatch login, ops that need an op slot but no cpu slot) would make this code clearer.

Reply:
@mkeilman has this been addressed?

Reply:
Not quite - see my last comment